mne/decoding/time_delaying_ridge.py

```diff
@@ -25,106 +40,100 @@ def _compute_corrs(X, y, smin, smax):
+                # linalg.toeplitz.
+                this_result = this_result[ij]
+                # However, we need to adjust for coeffs that are cut off,
+                # i.e. the non-zero delays should not have the same AC value
+                # as the zero-delay ones (because they actually have fewer
+                # coefficients).
+                #
+                # These adjustments also follow a Toeplitz structure, so we
+                # construct a matrix of what has been left off, compute their
+                # inner products, and remove them.
+                if edge_correction and smax > 0:
+                    tail = _toeplitz_dot(this_X[-1:-smax:-1, ch0],
+                                         this_X[-1:-smax:-1, ch1])
+                    if smin > 0:
+                        tail = tail[smin - 1:, smin - 1:]
+                    this_result[max(-smin + 1, 0):, max(-smin + 1, 0):] -= tail
+                if edge_correction and smin < 0:
+                    head = _toeplitz_dot(this_X[:-smin, ch0],
+                                         this_X[:-smin, ch1])[::-1, ::-1]
+                    if smax < 0:
+                        head = head[:smax, :smax]
+                    this_result[:-smin, :-smin] -= head
+
+                # Store the results in our output matrix
                 x_xt[ch0 * len_trf:(ch0 + 1) * len_trf,
                      ch1 * len_trf:(ch1 + 1) * len_trf] += this_result
                 if ch0 != ch1:
                     x_xt[ch1 * len_trf:(ch1 + 1) * len_trf,
                          ch0 * len_trf:(ch0 + 1) * len_trf] += this_result.T
+                count += 1
+                pb.update(count)
 
             # compute the crosscorrelations
-            cc_temp = np.fft.irfft(
-                y_fft * X_fft[:, ch0][:, np.newaxis].conj(), n_fft, axis=0)
+            cc_temp = cuda_dict['irfft'](
+                y_fft * X_fft_conj[:, slice(ch0, ch0 + 1)], n=n_fft, axis=0)
             if smin < 0 and smax >= 0:
                 x_y[:-smin, ch0] += cc_temp[smin:]
                 x_y[len_trf - smax:, ch0] += cc_temp[:smax]
             else:
                 x_y[:, ch0] += cc_temp[smin:smax]
+            count += 1
+            pb.update(count)
+    pb.done()
 
     x_y = np.reshape(x_y, (n_ch_x * len_trf, n_ch_y), order='F')
-    return x_xt, x_y, n_ch_x
+    return x_xt, x_y, n_ch_x, X_offset, y_offset
+
+
+def _toeplitz_dot(a, b):
+    """Create upper triangular Toeplitz matrices & compute the dot product."""
+    # This is equivalent to:
+    # a = linalg.toeplitz(a)
+    # b = linalg.toeplitz(b)
+    # a[np.triu_indices(len(a), 1)] = 0
+    # b[np.triu_indices(len(a), 1)] = 0
+    # out = np.dot(a.T, b)
+    assert a.shape == b.shape and a.ndim == 1
+    out = np.outer(a, b)
+    for ii in range(1, len(a)):
+        out[ii, ii:] += out[ii - 1, ii - 1:-1]
+        out[ii + 1:, ii] += out[ii:-1, ii - 1]
+    return out
```

Comment:
And this is faster than doing the numpy thing?

Reply:
Yeah